### ISSUE TITLE
Scores of train data

### ISSUE THREAD

**Issue by Pourbahman** [canjiali/PARADE#16]
Hi,

I want to give training data of Robust04 to your model and take scores like [https://zenodo.org/record/3974431/files/robust04.PARADE.runs.tar.gz](url) done for test data. Although your evaluated scores are for 3 epochs while in your paper is 36 epochs. So, it is better to ask how can I evaluate scores for both test and train data. Would you please guide me how to do it?

Thanks in advance,
Kind Regards

**Comment by canjiali** [comment]
please checkout https://github.com/canjiali/PARADE/blob/master/run_reranking.py#L416 for saving the scores.
please checkout https://github.com/canjiali/PARADE/blob/master/scripts/run.reranking.sh#L55 for model evaluation.

**Comment by Pourbahman** [comment]
Hi,

Thank you.

I add this lines of code to [https://github.com/canjiali/PARADE/blob/master/run_reranking.py#L416](url):

    eval_train_input_fn = input_fn_builder(
      dataset_path=[os.path.join(FLAGS.data_dir, "dataset_train.tfrecord")],
      max_num_segments_perdoc=FLAGS.max_num_segments_perdoc,
      max_seq_length=FLAGS.max_seq_length,
      is_training=False)

    trec_train_file = os.path.join(FLAGS.output_dir, "fold_{}_epoch_{}_bert_predictions_train.txt".format(FLAGS.fold, FLAGS.num_train_epochs))

    result_train = estimator.predict(input_fn=eval_train_input_fn,
                               yield_single_examples=True)
    results_train = []
    for item in result_train:
      results_train.append(
        (item["log_probs"], item["label_ids"]))
    log_probs_train, labels_train = zip(*results_train)
    log_probs_train = np.stack(log_probs_train).reshape(-1, 2)
    scores_train = log_probs_train[:, 1]
   
    result_info.write_result_from_score(
      rerank_topk=FLAGS.max_num_train_instance_perquery,
      scores=scores_train,
      qid_list=train_qid_list,
      relevance_dict=relevance_dict,
      write_path=trec_train_file,
      runid=FLAGS.aggregation_method
    )

Would you please check if it is correct, especially this line:
`rerank_topk=FLAGS.max_num_train_instance_perquery`

**Comment by canjiali** [comment]
Given that the dataset_train.tfrecord file is shuffled, you should generate another un-shuffled file for prediction.

**Comment by Pourbahman** [comment]
Hi,

Thanks! I think that it is enough to define the following code lines twice in main part of generate_data.py but one time is_eval=True would be set and dataset_train.tfrecord would be changed to dataset_train_eval.tfrecord. Do you agree?

```
# begin data convertion to TFrecord
  output_path = os.path.join(FLAGS.output_dir, "dataset_train.tfrecord")
  tf.logging.info("Writing data into {}".format(output_path))
  writer = tf.python_io.TFRecordWriter(output_path)
  convert_data_pointwise(
    writer=writer,
    tokenizer=tokenizer,
    qid_list=train_qid_list,
    relevance_dict=relevance_dict,
    corpus_dict=corpus_dict,
    query_dict=query_dict,
    is_eval=False
  )

```

**Comment by Pourbahman** [comment]
Hi,

Another question is:

I use 200G CPU RAM, Quadro 6000 GPU with 15G RAM, batch size = 10, epoch=1, fold=1 and it is 2 days that it is running. It is strange I think. Would you please tell me what your idea is?

I put the following in FLAGS.do_eval if condition and I think it is done after each epoch, so it should not be because of it. Also, I did no other changes!


> Hi,
> 
> Thank you.
> 
> I add this lines of code to [https://github.com/canjiali/PARADE/blob/master/run_reranking.py#L416](url):
> 
> ```
> eval_train_input_fn = input_fn_builder(
>   dataset_path=[os.path.join(FLAGS.data_dir, "dataset_train.tfrecord")],
>   max_num_segments_perdoc=FLAGS.max_num_segments_perdoc,
>   max_seq_length=FLAGS.max_seq_length,
>   is_training=False)
> 
> trec_train_file = os.path.join(FLAGS.output_dir, "fold_{}_epoch_{}_bert_predictions_train.txt".format(FLAGS.fold, FLAGS.num_train_epochs))
> 
> result_train = estimator.predict(input_fn=eval_train_input_fn,
>                            yield_single_examples=True)
> results_train = []
> for item in result_train:
>   results_train.append(
>     (item["log_probs"], item["label_ids"]))
> log_probs_train, labels_train = zip(*results_train)
> log_probs_train = np.stack(log_probs_train).reshape(-1, 2)
> scores_train = log_probs_train[:, 1]
> 
> result_info.write_result_from_score(
>   rerank_topk=FLAGS.max_num_train_instance_perquery,
>   scores=scores_train,
>   qid_list=train_qid_list,
>   relevance_dict=relevance_dict,
>   write_path=trec_train_file,
>   runid=FLAGS.aggregation_method
> )
> ```
> 
> Would you please check if it is correct, especially this line:
> `rerank_topk=FLAGS.max_num_train_instance_perquery`

**Comment by canjiali** [comment]
right

**Comment by Pourbahman** [comment]
> eval_train_input_fn = input_fn_builder(
>   dataset_path=[os.path.join(FLAGS.data_dir, "dataset_train.tfrecord")],
>   max_num_segments_perdoc=FLAGS.max_num_segments_perdoc,
>   max_seq_length=FLAGS.max_seq_length,
>   is_training=False)
> 
> trec_train_file = os.path.join(FLAGS.output_dir, "fold_{}_epoch_{}_bert_predictions_train.txt".format(FLAGS.fold, FLAGS.num_train_epochs))
> 
> result_train = estimator.predict(input_fn=eval_train_input_fn,
>                            yield_single_examples=True)
> results_train = []
> for item in result_train:
>   results_train.append(
>     (item["log_probs"], item["label_ids"]))
> log_probs_train, labels_train = zip(*results_train)
> log_probs_train = np.stack(log_probs_train).reshape(-1, 2)
> scores_train = log_probs_train[:, 1]
> 
> result_info.write_result_from_score(
>   rerank_topk=FLAGS.max_num_train_instance_perquery,
>   scores=scores_train,
>   qid_list=train_qid_list,
>   relevance_dict=relevance_dict,
>   write_path=trec_train_file,
>   runid=FLAGS.aggregation_method
> )

Hi,

Excuse me that I open the issue again but I have a question:
Would you please tell me since I want to have scores of train data, when I use unshuffled train data, train_qid_list that used in the above lines of code should be changed in https://github.com/canjiali/PARADE/blob/f330f12a0104c591d871aa54b3b0022dadaef512/run_reranking.py#L343

If yes, would you please tell me what I should change?

Thanks,
Zahra
